# Exercises for Lab 03 <font color=blue>Version 1.0 </font>

## Functions used for Tables:
Just like in Lab 02 Exercises, the bulk of this first part is to provide a reference for the functions you'll use. The tables here are taken from the datascience module, so all these table functions can only work if you import datascience (see the first line of code after this).

Remember that in a given table, lists make up the rows and arrays make up columns. Additionally, tables have separate indexes for rows and columns. An index of 0 could refer to the first column or first row (the header row is not given an index) depending on the function. However, if you're typing `existingTable[index]`, it will go for columns, not rows.

I'm using the same conventions as I did in Exercises 02, namely that all parts that you should change as befit the situation you are in are given capital letters, and those that are to stay the same are all-lowercase. However, the Table datatype naturally has a capital T, and as such, whenever you see a `Table` *by itself*, you shouldn't change that.

I'm also including the `[etc,etc]` feature, that states that the inputs for these methods can stretch on and on to infinity (assuming that the relevant pattern is still fulfilled). For example, the .with_columns method can add on infinitely many columns so long as the pattern of Header then Array is fulfilled.

### Functions
|Purpose|Function|
|--|--|
|Creating a New Table (from arrays)|`NewTable = Table().with_columns(ColumnHeader1, ListOrArray1, ColumnHeader2, ListOrArray2, [etc, etc])` (it's much recommended to make a table using arrays as columns rather than using lists as rows.)|
|Creating a New Table (from .csv file)|`NewTable = Table.read_table('FileName')` (if the file in question is not in the same folder as the Exercise/Lab you're working on, this won't work. You can technically add the file path at the beginning of the quoted FileName, but it's MUCH easier to just make sure the Notebook page you're working on is in the same folder as the file by default.)|
|Revealing the full table|`ExistingTable.show(NumberOfRows)` use empty parentheses if you want all rows|
|Restricting a Table to certain parameters|`ExistingTable.where(ColumnHeader, are.Condition(Parameter))` (check below for conditions/parameters)|
|Selecting  row in Table format|`ExistingTable.take(IndexOfRow1, IndexOfRow2, [etc, etc])` (you can use functions to create arrays to put here for the row indices)|
|Selecting a column in Table format|`ExistingTable.select('ColumnHeader1' OR IndexOfColumn1, 'ColumnHeader2' OR IndexOfColumn2, [etc, etc])` (stick to one type; either all headers or all indices.) OR `existingTable[IndexOfColumn]`|
|Selecting a column in array format|`ExistingTable.column('ColumnHeader1' OR IndexOfColumn1, 'ColumnHeader2' OR IndexOfColumn2, [etc, etc])`|
|Adding one row to a Table|`ExistingTable = ExistingTable.with_row(NewRowAsList)`|
|Adding one column to a Table|`ExistingTable = ExistingTable.with_column(ColumnHeader, ListOrArray)`|
|Adding multiple rows to a Table|`ExistingTable = ExistingTable.with_rows(NewRowAsList1, NewRowAsList2, [etc, etc])`|
|Adding multiple columns to a Table|`ExistingTable = ExistingTable.with_columns(ColumnHeader1, ListOrArray1, ColumnHeader2, ListOrArray2, [etc, etc])`|
|Sorting the rows|`ExistingTable.sort(ColumnHeader, descending = True/False)` (descending is automatically set to false; put it to True if you want highest values first)|
|Getting the number of rows/columns|`ExistingTable.num_rows` OR `ExistingTable.num_columns`|
|Getting rid of certain columns|`ExistingTable.drop(UnwantedColumnHeader1, UnwantedColumnHeader2, [etc, etc])`|

### Conditions
|Purpose|Function (in np.where)|
|--|--|
|where target values are equal to the Target Element|`ExistingTable.where(ColumnHeader, are.equal_to(TargetElement))` OR `ExistingTable.where(ColumnHeader, TargetElement)`|
|where target values are NOT equal to the Target Element|`ExistingTable.where(ColumnHeader, are.not_equal_to(TargetElement))`|
|where target values are less than the Target Element|`ExistingTable.where(ColumnHeader, are.below(TargetElement))`|
|where target values are more than the Target Element|`ExistingTable.where(columnHeader, are.above(TargetElement))`|
|where target values are between two Target Elements|`ExistingTable.where(ColumnHeader, are.between(Element1, Element2))` (this includes values equal to Element1, but NOT those equal to Element2)|
|where target values contain the Target Element|`ExistingTable.where(ColumnHeader, are.containing(TargetElement))`|


One key detail is that you need to keep track of indexes. Taking the index of a table once gives you a column, and taking the index twice gives you a cell. For example, `existingTable[5][2]` goes to the table's 6th column (remember, indexes start counting at 0, not 1) and gets the cell of that column that is in the 3rd row.

Before we can even hope to work with tables, we must import the relevant libraries. These relevant libraries are the datascience module, which were will almost always import without its name.

In [1]:
from datascience import *
import numpy as np # this library is also used in these Exercises.

## Exercise 1: 
A central part of data manipulation is, well, the manipulation of data (well done Einstein). This primarily constitutes the use of FOM (Functions, Operations, and/or Methods) to show another aspect of data or to express existing data in a different format.

Why do we want to manipulate data?
* To remake it in a format that is more easily usable by the computer itself (Remember, your capabilities in analyzing data are determined at least partially by the capabilities of Jupyter itself!)
* To conduct statistical tests and other advanced actions (spoilers for Lab 05 and onwards)

The principal way to manipulate data is to extract it from the Table setting. While it may seem that Tables are the all-important focus of this class, don't fall for that illusion. **Tables are only a convenient way to organize and understand data. By themselves, they are incapable of actual analysis.** The true leaders here are the lists and arrays. I'll leave it to you to find the exact method that converts tables to arrays in the above table.

### 1.1: Title Length
Create the imdb Table from the file named 'imdb.csv', and extract the Title column as an array. 

Use the relevant command, and the correct format, to get a list that contains the number of characters in each title. For example, the movie title 'León' would beget a value of 4 (it has 4 characters). Review the Exercises 02 if you do not know.

Append this column to the imdb table. Give it any column header you want. :']

What are the five movies with the longest names?

### 1.2: Multistep process
Add a column to the imdb table that shows how many a's that movie title has. For example, the movie title 'All About Eve' would have a value of 2.

I'll walk you through this one, since it does a good job of showing how multistep coding procedures work in regards to data manipulation.

In [ ]:
imdb

Step 1: You may have noticed that the letter A sometimes appears as uppercase (A) and other times as lowercase (a). Python treats these two as totally distinct characters (like apples and oranges). Extract the 'Title' column as an array again (or use the same variable if you already defined it in Exercise 1.1) and use the method `String.lower()` to convert all characters to lowercase. Hmm, how do you apply a method to all elements in a list?

In [ ]:
lowercase_titles = ...

Step 2: Now comes the heart of the problem, that is, how to make Python count all the a's in a title. There are many ways to go about this, and you are welcome to create your own solution here. However, one of the simplest I know revolves around the `String.split(splitBoundary)` method, which was introduced in the Exercises for Lab 02. Refer back there (particularly to the 2.3 and 2.4 exercise problems) or to your own toolbox (that you *should* create) for how to use that method.

Note that, in the example below, for each instance of A, the method effectively cleaves the string at that location. Thus, you can expect that a word with one A begets two lists when put in the .split method, two A's begets three lists, and so forth.

Your lightbulb may already be alight about where I'm going with this, but I'll take it slow in case you're still confused. Use .split to get the various cleavings of each title. This will be a list of lists.

In [ ]:
# examples!
print('chameleon'.split('a')) # one a begets two fragments
print('bargain'.split('a')) # two a begets three fragments
print('avalanche'.split('a')) # three a begets four fragments (note that the .split includes a separate fragment even though a is the first letter!)

# in a list:
example_words = make_array('chameleon', 'bargain', 'avalanche') # this represents the array of titles
split_list = [x.split('a') for x in example_words] # this is a list of lists
split_list

In [ ]:
# Use the examples to get the various fragments of the titles in your title column array
...

Step 3: Use another function in Exercises 02 to get the length of each sub-list in the list. Adjust the values so that they match the number of a's. (Remember, a list of fragments of length 3 came from a title that had TWO a's, not three!)

In [ ]:
# example! A function on all sublists of a list. This function is NOT what you want for step 3, however.
[x[0] for x in split_list] # this takes the first term of all the sublists.

Step 4: Append the column to the imdb table.

What are the five movies with the most A's in the title?

### 1.3: Date Formatting
Now time for some actually-real-world applications. A very common sight in Tables and Data Statistics in general is the reformatting of dates to render them more code-oriented. After all, Python has no idea what the slashes in '7/31/2025' mean, nor does it comprehend the colons in '6:30:25'. To be able to work with times and dates, we need to reexpress these values in terms of a single unit (e.g. hours) instead of multiple units (e.g. hours : minutes : seconds).

Consider the times in the following table. Our task in this exercise is to convert them from hh:mm:ss to some amount of hours past midnight. For example, a time of 11:30:00 gets a value of 11.5 hours past midnight, and 7:25:52 gets a value of 7.431 hours past midnight. In essence, we are doing the following: (using 15:35:20 as an example)
1. Getting the seconds in terms of a decimal portion of a minute (e.g. 20 seconds is 1/3 of a minute or 0.33 minutes)
2. Adding that to the minute count (35+0.33 = 35.33 minutes)
3. Expressing that minute count in terms of a decimal portion of an hour (35.33 minutes becomes 0.589 hours)
4. Adding that to the hours (15+0.589 = 15.589 hours)

In [ ]:
unformatted = ["4:50:31", "7:12:45", "9:38:09", "11:27:56", "13:03:22", "15:41:10", "17:19:44", "20:02:37", "22:46:05", "00:15:28"]

times = Table().with_columns('hh:mm:ss times', unformatted)
times # this is typically what these times will look like within a table. 
# Of course, other data will also be there, but for the sake of simplicity, this is not included.

When having to format a whole list, it's a good idea to focus solely on one singular value, and then extrapolate to the entire thing. Try working with just this first value below.

In [ ]:
test_time = times[0][0]
test_time

Here are some guiding questions:
* How do you 'unpack' the values in the data format?
* Are these unpacked values able to be worked with numerically? What do you have to do to make them so?
* Once you have have worked it out and gotten the singular test_time to work, what do you have to do to apply this to the whole table column?
* Remember, what first step do you have to do to make data within tables open to analysis?

Once you've solved it, append the formatted times as another column to the table.

## Exercise 2:
A crucial part of data science is being able to translate abstract questions into code lines that will give you the answer to the question. This type of translation is akin to mathematical calculations or finding a route to drive to a certain place. You've already seen such questions in the lab itself, but given how important they are to your projects, you will need plenty more practice for this. The main strategy is to know what your answer should look like and how to get there, and therefore to design a series of steps to get from the entire table to your final destination.

### 2.1: World Populations
These questions regard the dataset 'world_populations.csv'. I've done some prep work below to include the years.

In [ ]:
pops = Table.read_table('world_population.csv')
pops = pops.with_column('Year', np.arange(1950, 2016))
pops

How many years did the world spend with a population between 4 and 5 billion?

Has the world population grown a larger amount from 1950 to 2000 or from 2000 to 2016? By how much are these amounts different?

### 2.2: Farmers Markets
These questions regard the dataset 'farmers_markets.csv'.

In [ ]:
farmers_markets = Table().read_table('farmers_markets.csv')
farmers_markets = farmers_markets.drop('FMID','Website','Facebook','Twitter','Youtube','OtherMedia','Season1Date','Season1Time','Season2Date','Season2Time','Season3Date','Season3Time','Season4Date','Season4Time')
farmers_markets

What percentage of these markets have 'Farmers' in their name?

What markets lie within 1 degree, latitude and longitude, of the Temple University's SERC building (the coordinates are -75.153, 39.982)? For reference, one degree of latitude/longitude is equal to 69 miles. (nice)

What markets sell Meat and Tofu but no vegetables?

## Exercise 3:
### 3.1:
In this section, we will introduce to ourselves a very useful function whose name I will not spoil yet. This involves the use of the farmers_markets table. If you look at the bunch of columns at the far right end, most of them show whether or not a given market sells a particular good. A 'Y' means that the market does sell it, and 'N' means that it doesn't.

Suppose you pick the Cheese column (because I like cheese) that has these Ys and Ns and want to know how many of them have Ys (how many sell cheese) and how many of them have Ns (how many don't sell cheese). (Note that Ys and Ns are the only inputs in the column.) How do you do this?

Step 1: Define the `Y_cheese` variable as the farmers_markets table containing only the rows with 'Y' in the Cheese column, and the `N_cheese` variable as the table containing only the rows with 'N' in the Cheese column.

Step 2: Get the number of rows in the Y_cheese and N_cheese, and define them as `Y_rows` and `N_rows`.

Note that what you've gotten right now, in Y_rows and N_rows, is the number of markets that have cheese (number of rows that have 'Y') and don't have cheese (number of rows that have 'N'). Now, assuming you've done your duty in properly defining these variables, I'll package this information in a format that is much more pleasant to your eyes.

In [ ]:
# don't change this cell!
inputs = ['Y', 'N']
number_rows = [Y_rows, N_rows]
summary_cheese = Table().with_columns('Cheese', inputs, 'count', number_rows)
summary_cheese = summary_cheese.sort('Cheese')
summary_cheese

This is the essence of our mystery function: for each unique element in a given column, it will tell you how many rows in the table have that input for that column. Here is that mystery function below.

In [ ]:
farmers_markets.group('Cheese')

### 3.2: Practice with `.group`
Which state has the most markets?

Going back to the imdb table, which decade had the most movies?

In [ ]:
imdb

What's the maximum number of A's a movie had? Create the .group table and assign it the variable `A_stats`.

Note that, since the .group method gives you a table, you can modify that table in itself. Use this fact to add on to `A_stats` another column that tells you what percentage of the total movies had that number of A's. (e.g. What percent had zero A's? What percent had one A? and so forth)

### 3.3: `.group` with a function
When paired with a function as the second argument, the resultant table is very different. Let's look back at the imdb table, and apply the .group method and the np.average function to the 'Decade' column. It does filter the table according to each unique input of Decade, but instead of taking the number of rows for each input in Decade, it looks at the other columns besides Decade and takes their average (applies the specified function).

In [ ]:
imdb.group('Decade', np.average)

Let's take the first row as an example. 

What it tells us is that, for all the 1920 movies in the table...
* They got 64854 votes on average
* They had an average rating of 8.25
* The average year in which they were released was 1924 (ignoring decimals)
* They had an average of 0.25 A's in their name.

And this goes for all the movies released in the decade of 1920. The 1930 movies' average stats are in the second row, the 40s' stats in the third, and so forth. 

Note that the Title column is blank because you can't take the average of non-numerical values.

**What functions can you use in `.group`?**

Any function works SO LONG AS its only input is a list or array. Methods do not work, as the .group assumes that this only input is in the parentheses following the function name, and the method's input is before it.

Assume that you have already done 1.2 and thus have the 'Number of As' column at your disposal. If you're stuck on some of these, you are more than welcome to create extra code lines to study the relevant data table or make your steps a bit more coherent.

What's the total number of A's in movies of the 1940s?

What decade had the longest-titled movie names?

What's the westernmost market that sells Eggs?

Now here's something that's much more in line with 1.2. Do countries ending in '-ia' generally have a higher epi score than those that do not? You can reasonably expect that all countries that have an 'ia' will have it at the very end (that is, there is no difference between having it at the end and generally containing 'ia' at all), so you can just use the .split method that was introduced in 1.2.

You may want to adjust it such that those with 'ia' have a value of 1 and those without have a value of 0. 

## Exercise 4: Pivot
A function that is similar to .group is .pivot, where the resultant table denotes how many rows have TWO indicated elements in TWO columns. Take a look at the table formed by this following cell.

In [4]:
imdb.pivot('Rating', 'Decade')

Decade | 8.0  | 8.1  | 8.2  | 8.3  | 8.4  | 8.5  | 8.6  | 8.7  | 8.8  | 8.9  | 9.0  | 9.2
1920   | 0    | 0    | 2    | 2    | 0    | 0    | 0    | 0    | 0    | 0    | 0    | 0
1930   | 0    | 3    | 1    | 0    | 1    | 2    | 0    | 0    | 0    | 0    | 0    | 0
1940   | 2    | 3    | 1    | 4    | 2    | 1    | 1    | 0    | 0    | 0    | 0    | 0
1950   | 6    | 7    | 5    | 5    | 3    | 2    | 0    | 1    | 0    | 1    | 0    | 0
1960   | 6    | 4    | 2    | 6    | 0    | 2    | 1    | 0    | 0    | 1    | 0    | 0
1970   | 5    | 5    | 1    | 4    | 0    | 2    | 0    | 2    | 0    | 0    | 1    | 1
1980   | 5    | 6    | 6    | 4    | 7    | 2    | 0    | 1    | 0    | 0    | 0    | 0
1990   | 9    | 4    | 5    | 5    | 3    | 4    | 5    | 3    | 1    | 2    | 0    | 1
2000   | 8    | 14   | 8    | 5    | 4    | 6    | 1    | 1    | 1    | 2    | 0    | 0
2010   | 9    | 7    | 3    | 3    | 2    | 3    | 1    | 1    | 0    | 0    | 0    | 0

Looking at the first row, it tells us that, among the 1920 movies, 2 of them scored an 8.2. That is, there were two rows that simultaneously had a value of 1920 for the decade and 8.2 for the rating. 

Note the input order; the column elements on the headers correspond to the FIRST input, and those on the side axis correspond to the SECOND input. Remember, columns are much more workable than rows when it comes to data manipulation.

### 4.1: 
Which state has the largest *proportion* of markets that sell seafood? Before you start, think about this question and come up with a hypothesis.

Hypothesis: *type here*

Make a pivot table and add necessary adjustments to display the proportions. Think back to problems like 1.1 and 1.3 for the general steps (but how do you get the proportions using your given data?).

### 4.2:
Do markets on a Local gvernment building grounds generally accept credit cards more often than those that are on a private business parking lot?

## Conclusion
And that's the end of Exercises 03! I admit, this was a beefy one. Please provide feedback and any avenues for improvement in the cell below.

Unfortunately, I didn't put any secret codes in this Exercise 03. (Gotta make all five points really worth your while!)